File: clients/ios-macos/PassGFW/Sources/PassGFW/NetworkClient.swift
import Foundation

/// HTTP Response
struct HTTPResponse {
    let success: Bool
    let statusCode: Int
    let body: String
    let error: String?
}

/// Network Client for HTTP requests
class NetworkClient {
    private let timeout: TimeInterval
    
    init(timeout: TimeInterval = Config.requestTimeout) {
        self.timeout = timeout
    }
    
    /// POST request
    func post(url: String, jsonBody: String) async -> HTTPResponse {
        guard let requestURL = URL(string: url) else {
            return HTTPResponse(success: false, statusCode: 0, body: "", error: "Invalid URL")
        }
        
        var request = URLRequest(url: requestURL)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("PassGFW/1.0 Swift", forHTTPHeaderField: "User-Agent")
        request.timeoutInterval = timeout
        request.httpBody = jsonBody.data(using: .utf8)
        
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                return HTTPResponse(success: false, statusCode: 0, body: "", error: "Invalid response")
            }
            
            let body = String(data: data, encoding: .utf8) ?? ""
            let success = (200...299).contains(httpResponse.statusCode)
            
            return HTTPResponse(
                success: success,
                statusCode: httpResponse.statusCode,
                body: body,
                error: success ? nil : "HTTP \(httpResponse.statusCode)"
            )
        } catch {
            return HTTPResponse(success: false, statusCode: 0, body: "", error: error.localizedDescription)
        }
    }
    
    /// GET request
    func get(url: String) async -> HTTPResponse {
        guard let requestURL = URL(string: url) else {
            return HTTPResponse(success: false, statusCode: 0, body: "", error: "Invalid URL")
        }
        
        var request = URLRequest(url: requestURL)
        request.httpMethod = "GET"
        request.setValue("PassGFW/1.0 Swift", forHTTPHeaderField: "User-Agent")
        request.timeoutInterval = timeout
        
        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            
            guard let httpResponse = response as? HTTPURLResponse else {
                return HTTPResponse(success: false, statusCode: 0, body: "", error: "Invalid response")
            }
            
            let body = String(data: data, encoding: .utf8) ?? ""
            let success = (200...299).contains(httpResponse.statusCode)
            
            return HTTPResponse(
                success: success,
                statusCode: httpResponse.statusCode,
                body: body,
                error: success ? nil : "HTTP \(httpResponse.statusCode)"
            )
        } catch {
            return HTTPResponse(success: false, statusCode: 0, body: "", error: error.localizedDescription)
        }
    }
}

File: clients/android/passgfw/src/main/kotlin/com/passgfw/NetworkClient.kt
package com.passgfw

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.util.concurrent.TimeUnit

/**
 * HTTP Response
 */
data class HTTPResponse(
    val success: Boolean,
    val statusCode: Int,
    val body: String,
    val error: String?
)

/**
 * Network Client for HTTP requests
 */
class NetworkClient(private val timeout: Long = Config.REQUEST_TIMEOUT) {
    private val client = OkHttpClient.Builder()
        .connectTimeout(timeout, TimeUnit.MILLISECONDS)
        .readTimeout(timeout, TimeUnit.MILLISECONDS)
        .writeTimeout(timeout, TimeUnit.MILLISECONDS)
        .build()
    
    private val jsonMediaType = "application/json; charset=utf-8".toMediaType()
    
    /**
     * POST request
     */
    fun post(url: String, jsonBody: String): HTTPResponse {
        return try {
            val request = Request.Builder()
                .url(url)
                .post(jsonBody.toRequestBody(jsonMediaType))
                .addHeader("Content-Type", "application/json")
                .addHeader("User-Agent", "PassGFW/1.0 Kotlin")
                .build()
            
            client.newCall(request).execute().use { response ->
                HTTPResponse(
                    success = response.isSuccessful,
                    statusCode = response.code,
                    body = response.body?.string() ?: "",
                    error = if (response.isSuccessful) null else "HTTP ${response.code}"
                )
            }
        } catch (e: Exception) {
            HTTPResponse(false, 0, "", e.message)
        }
    }
    
    /**
     * GET request
     */
    fun get(url: String): HTTPResponse {
        return try {
            val request = Request.Builder()
                .url(url)
                .get()
                .addHeader("User-Agent", "PassGFW/1.0 Kotlin")
                .build()
            
            client.newCall(request).execute().use { response ->
                HTTPResponse(
                    success = response.isSuccessful,
                    statusCode = response.code,
                    body = response.body?.string() ?: "",
                    error = if (response.isSuccessful) null else "HTTP ${response.code}"
                )
            }
        } catch (e: Exception) {
            HTTPResponse(false, 0, "", e.message)
        }
    }
}

